### DISCUSSION TITLE
How to update volume?

### DISCUSSION THREAD

**Discussion by KestutisMa** [pyvista/pyvista#5116] [Q&A]
Dear PyVista team,

I am trying to update Volume plot, when scalar data changes on initial ImageData. I am using PyQt with PyVista.

**Problem**: Plot is updated when same `ImageData` instance is used with `add_mesh`, but does not update when similar plot is created with `add_volume`. In that case only initial data is plotted and plot is not updating when using `self.vox['vm']=...`.

```
self.vox = ImageData(...)
...
self.vol = self.plotterVoltage.add_volume( self.vox, scalars='vm') # not updating
self.vol = self.plotterVoltage.add_mesh( self.vox, scalars='vm') # updating

...
self.vox['vm'] = np.random.rand(self.vox['vm'].shape[0]) # update plot with some random data
....
```

I did good bit of digging examples and documentation without any luck.

How to update Volume plot?

Best,
Kestutis

**Comment by neale** [comment]
I second this. Looks like the `update_scalars()` function inside the plotter is expecting a mesh. 

When calling `p.update_scalars(values)` on a volume I get:
```
if scalars.shape[0] == mesh.GetNumberOfPoints():
AttributeError: 'NoneType' object has no attribute 'GetNumberOfPoints'
```

I naively tried to replace the `mapper.dataset` inside the plotter, but that didnt work. 

Example:

```python
def render(volume, grid=None, p=None):
    """render volume with pyvista"""

    grid = grid if grid is not None else pv.UniformGrid(dimensions=volume.shape[:-1])
    opacity = transfer_function(volume)
    rgba = np.concatenate((volume, opacity), -1).reshape(-1, 4).astype(np.uint8)
    if p is None:
        p = pv.Plotter(notebook=False,
                       multi_samples=8,
                       line_smoothing=True,
                       polygon_smoothing=True,)
        p.add_volume(grid,
                    scalars=rgba,
                    ambient=1.0,
                    mapper="smart",
                    diffuse=.5,
                    blending='composite',
                    shade=False)
        p.show(interactive_update=True)
    else:
        p.update_scalars(rgba) # error
        p.update()

    return grid, p
```

Let me know if you need a fully workable example to reproduce this.

**Reply by banesullivan** [reply]
FYI, I just deprecated `update_scalars` in #5158

**Comment by banesullivan** [comment]
This is an issue specific to `add_volume()`. I'm pretty sure we create a copy/"view" of the `ImageData` in this method to do some alterations to the data to make it "volume-rendering-ready":

https://github.com/pyvista/pyvista/blob/5e1b0ec2b8681c6b04b857ee30e670fdf46b97e6/pyvista/plotting/plotter.py#L4035-L4038

I think there is a way around this... I'll investigate and try to provide a solution... stay tuned.

**Reply by banesullivan** [reply]
Shucks, we need to implement `vtkAlgorithm` support for `add_volume()` first to make this work (ref #3318)

This is tracked in #4626

**Reply by neale** [reply]
As a side note. volume rendering really seems to be a second class citizen in pyvista, compared to meshes. I get the feeling that this is because vtk doesn't handle volumes well, is that true?

**Reply by banesullivan** [reply]
I can understand why you think that and it's definitely not intentional. Volume rending is well handled in VTK, but PyVista has primarily focused on mesh rendering due to it being the primary thing our developers and users want to do. There's really no hurdles to making volume rendering as seamless as mesh rendering here besides funding/developer time.

I think improving volume rendering would be a great thing for us to pursue a small development grant for, but unfortunately I do not have the availability to work on that at this time

**Reply by neale** [reply]
I see, I guess I only see it that way since I only interact with pyvista for volume rendering. 
I appreciate the work that goes into this.

**Comment by Datamance** [comment]
I may be a little late to the game on this one, but the only dependable way I've found is to grab it off the mapper _after_ calling `add_volume`

```python
grid = pv.ImageData()
...
volume = plotter.add_volume(
    grid,
    opacity=opacities,
    cmap="bwr",
)
...
grid: pv.ImageData = self.plotter.mapper.dataset
```

**Reply by neale** [reply]
How do you use this in a loop to update the plotter? 
Are you continuously calling `add_volume`, or updating `plotter.mapper.dataset.point_data["Data"]`

**Comment by Datamance** [comment]
Also, for some reason (that I couldn't find in the docs), the active scalars name is changed to `"Data"` so make sure to account for that in the setting
```python
grid.point_data["Data"] = normalized.astype(np.uint8).ravel(order="F")
```
